ethereum/geth_client.go

```diff
@@ -59,217 +58,251 @@ type Config interface {
 			err = result[1].Interface().(error)
 		}
 
-		if err != nil {
-			if (err.Error() == TransactionUnderpriced) || (err.Error() == NonceTooLow) {
-				log.Warningf("Concurrent transaction identified, trying again [%d/%d]\n", current, maxTries)
-				time.Sleep(gethClient.config.GetEthereumIntervalRetry())
-			} else {
-				done = true
-			}
-		} else {
-			done = true
+		if err == nil {
+			return tx, nil
 		}
+
+		if (err.Error() == transactionUnderpriced) || (err.Error() == nonceTooLow) {
+			log.Warningf("Concurrent transaction identified, trying again [%d/%d]\n", current, maxTries)
+			time.Sleep(gc.config.GetEthereumIntervalRetry())
+			continue
+		}
+
+		return nil, err
 	}
 
-	return
 }
 
-func (gethClient *GethClient) incrementNonce(opts *bind.TransactOpts) (err error) {
-	if !gethClient.config.GetTxPoolAccessEnabled() {
-		log.Warningf("Ethereum Client doesn't support txpool API, may cause concurrency issues.")
-		return
+// noncer defines functions to get the next nonce
+type noncer interface {
+	PendingNonceAt(ctx context.Context, account common.Address) (uint64, error)
+}
+
+// callContexter defines functions to get CallContext
+type callContexter interface {
+	CallContext(ctx context.Context, result interface{}, method string, args ...interface{}) error
+}
+
+// incrementNonce updates the opts.Nonce to next valid nonce
+// We pick the current nonce by getting latest transactions included in the blocks including pending blocks
+// then we check the txpool to see if there any new transactions from the address that are not included in any block
+// If there are no pending transactions in txpool, we use the current nonce + 1
+// else we increment the greater of current nonce or the nonce derived from txpool
+func incrementNonce(opts *bind.TransactOpts, txpoolAccessEnabled bool, noncer noncer, cc callContexter) error {
+	// check if the txpool access is enabled
+	if !txpoolAccessEnabled {
+		log.Warningf("Ethereum Client doesn't support txpool API, may cause transactions failures")
+		opts.Nonce = nil
+		return nil
 	}
 
-	var res map[string]map[string]map[string][]string
-	// Important to not create a DeadLock if network latency
-	txctx, _ := context.WithTimeout(context.Background(), GetDefaultContextTimeout())
-	gc.GetRpcClient().CallContext(txctx, &res, "txpool_inspect")
+	ctx, cancel := context.WithTimeout(context.Background(), GetDefaultContextTimeout())
+	defer cancel()
 
-	if len(res["pending"][opts.From.Hex()]) > 0 {
-		chainNonce, err := gc.GetClient().PendingNonceAt(txctx, opts.From)
-		if err != nil {
-			log.Errorf("Error found when getting the Account Chain Nonce [%v]\n", err)
-			return err
-		}
-		CalculateIncrement(chainNonce, res, opts)
+	// get current nonce
+	n, err := noncer.PendingNonceAt(ctx, opts.From)
+	if err != nil {
+		return fmt.Errorf("failed to get chain nonce for %s: %v", opts.From.String(), err)
 	}
 
-	return
-}
+	// set the nonce
+	opts.Nonce = new(big.Int).Add(new(big.Int).SetUint64(n), big.NewInt(1))
```

Comment:
I rather not increment by default, this could cause problems. I do not think we need this here, as it should only be incremented when there are many pending txs.

Reply:
We are setting the nonce here but we will still look up for any pending txns. If we find the any pending txns, we will use that